pkg/authorization/api/validation/validation.go

```diff
@@ -8,48 +8,126 @@ import (
 	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
 )
 
-// ValidatePolicyBinding tests required fields for a PolicyBinding.
-func ValidatePolicyBinding(policyBinding *authorizationapi.PolicyBinding) fielderrors.ValidationErrorList {
+func ValidatePolicyName(name string, prefix bool) (bool, string) {
+	if name != authorizationapi.PolicyName {
+		return false, "name must be " + authorizationapi.PolicyName
+	}
+
+	return true, ""
+}
+
+func ValidatePolicy(policy *authorizationapi.Policy) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
+	allErrs = append(allErrs, validation.ValidateObjectMeta(&policy.ObjectMeta, true, ValidatePolicyName).Prefix("metadata")...)
+
+	for roleKey, role := range policy.Roles {
+		if roleKey != role.Name {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("roles."+roleKey+".name", policy, "must be "+roleKey))
+		}
+
+		allErrs = append(allErrs, ValidateRole(&role).Prefix("roles."+roleKey)...)
+	}
+
+	return allErrs
+}
 
-	if len(policyBinding.Name) == 0 {
-		allErrs = append(allErrs, fielderrors.NewFieldRequired("name"))
+func ValidatePolicyUpdate(oldPolicy *authorizationapi.Policy, policy *authorizationapi.Policy) fielderrors.ValidationErrorList {
+	allErrs := ValidatePolicy(policy)
+	allErrs = append(allErrs, validation.ValidateObjectMetaUpdate(&oldPolicy.ObjectMeta, &policy.ObjectMeta).Prefix("metadata")...)
+
+	return allErrs
+}
+
+func ValidatePolicyBindingName(policyRefNamespace string) validation.ValidateNameFunc {
+	return func(name string, prefix bool) (bool, string) {
+		if name != policyRefNamespace {
+			return false, "name must be " + policyRefNamespace
+		}
+
+		return true, ""
 	}
+}
+
+// ValidatePolicyBinding tests required fields for a PolicyBinding.
+func ValidatePolicyBinding(policyBinding *authorizationapi.PolicyBinding) fielderrors.ValidationErrorList {
+	allErrs := fielderrors.ValidationErrorList{}
+	allErrs = append(allErrs, validation.ValidateObjectMeta(&policyBinding.ObjectMeta, true, ValidatePolicyBindingName(policyBinding.PolicyRef.Namespace)).Prefix("metadata")...)
 
 	if len(policyBinding.PolicyRef.Namespace) == 0 {
 		allErrs = append(allErrs, fielderrors.NewFieldRequired("policyRef.Namespace"))
 	}
 
-	allErrs = append(allErrs, validation.ValidateLabels(policyBinding.Labels, "labels")...)
+	for roleBindingKey, roleBinding := range policyBinding.RoleBindings {
+		if roleBinding.RoleRef.Namespace != policyBinding.PolicyRef.Namespace {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("rolebindings."+roleBindingKey+".roleRef.namespace", policyBinding, "must be "+policyBinding.PolicyRef.Namespace))
+		}
+
+		if roleBindingKey != roleBinding.Name {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("rolebindings."+roleBindingKey+".name", policyBinding, "must be "+roleBindingKey))
+		}
+
+		allErrs = append(allErrs, ValidateRoleBinding(&roleBinding).Prefix("rolebindings."+roleBindingKey)...)
```

Comment:
you're supposed to use the internal names, and then eventually we'll convert them on the way out in serializers.